### ISSUE TITLE
[1.12.2] ItemSnow behaviour is completely broken for modded blocks

### ISSUE THREAD

**Issue by Sollace** [MinecraftForge/MinecraftForge#5616]
Minecraft Version: 1.12.2
Forge Version: 14.23.2.2654

I _highly_ doubt anyone's going to resolve this issue before Minecraft 1.15, but everyone I've spoken to about it says they have to deal with fixing the same issue on their projects and that it's worth a shot.

So here's the situation:
Say you have a mod where you want to add a custom block with the same behaviour as minecraft:snow_layer (a block with 8 layers that increases every time you right-click with that block's item).

The obvious solution is to extend BlockSnow and use the ItemSnow as the item, expecting it should just work.

The block works fine, but here's the issue: The item doesn't. Every time you try to place the block down, you can right-click it once to get to 2 layers, and then it stops increasing.

What's expected is that it would continue to increase to a full block, as per vanilla behaviour.

After some debugging and looking through the code, I eventuallt found the issue:
ItemSnow.java

    public boolean canPlaceBlockOnSide(World world, BlockPos pos, EnumFacing side, EntityPlayer player, ItemStack stack)
    {
        IBlockState state = world.getBlockState(pos);
        return (state.getBlock() != net.minecraft.init.Blocks.SNOW_LAYER || ((Integer)state.getValue(BlockSnow.LAYERS)) > 7) ? super.canPlaceBlockOnSide(world, pos, side, player, stack) : true;
    }

_Specifically_, this part:

    state.getBlock() != net.minecraft.init.Blocks.SNOW_LAYER || 

This is wrong, and judging by how the Blocks class is fully qualified, I'm led to believe this is not a vanilla issue, but some change that's been brought in with forge's patches.

What I ended up having to do is extend _that_ class, and specifically undo the change (as best I can):

        @Override
    public boolean canPlaceBlockOnSide(World world, BlockPos pos, EnumFacing side, EntityPlayer player, ItemStack stack) {
        IBlockState state = world.getBlockState(pos);

        // Check explicitly for the maximum layers.
        // Without this layered blocks get stuck on the second layer because of a change introduced with Forge's patches.
        if (state.getBlock() instanceof BlockSnow && state.getValue(BlockSnow.LAYERS) < 8) {
            return true;
        }

        // return (state.getBlock() != net.minecraft.init.Blocks.SNOW_LAYER || ((Integer)state.getValue(BlockSnow.LAYERS)) > 7) ? super.canPlaceBlockOnSide(world, pos, side, player, stack) : true;
        return super.canPlaceBlockOnSide(world, pos, side, player, stack);
    }

And that works as expected.

You can draw your own conclusions from this, but my belief is that forge **should not have touched that method in the first place.**

**Comment by Sollace** [comment]
Also an extra snippet from discord where I explain to the best of my understanding why that code causes ItemSnow to break the way it is:

It wasn't clear to me either.
Here is what that code is actually doing:
    if (state.getBlock() != Blocks.SNOW_LAYER || state.getValue(BlockSnow.LAYERS) > 7) {
      return super.canPlaceBlockOnSide(world, pos, side, player, stack);
   }
   return true;

The super method ends up returning false, because it's the generic checks for ItemBlock.
And those will generally end up offsetting to the block above it.
Which can't have a block placed because the snow layer isn't at full height.

**Comment by bs2609** [comment]
I'm a little confused here - the only difference between the current Forge version and your version is a change from `state.getBlock() == Blocks.SNOW_LAYER` to `state.getBlock() instanceof BlockSnow`.
Is that what you are intending to say the problem is?

As for the patch, you can look at the history to see where it was introduced:
b4779be2f639dee55738db502531ac0f4a454895, to fix #1534.

As that dates back to 1.8, it is possible it's no longer needed due to a fix in vanilla in the meantime, but it seems unlikely that any vanilla code will have been written with modded blocks in mind.

**Comment by tterrag1098** [comment]
This change is still needed. Forge patches World.mayPlace to use the block's `isReplaceable` method rather than the material. This requires the check for isReplaceable to be added to ItemBlock [here](https://github.com/MinecraftForge/MinecraftForge/blame/1.12.x/patches/minecraft/net/minecraft/item/ItemBlock.java.patch#L39). Which in turn requires ItemSnow to specially override `canPlaceBlockOnSide` to handle the odd interaction logic.

TL;DR: It's intended that you need to override this function.

**Comment by Sollace** [comment]
> 
> 
> I'm a little confused here - the only difference between the current Forge version and your version is a change from `state.getBlock() == Blocks.SNOW_LAYER` to `state.getBlock() instanceof BlockSnow`.
> Is that what you are intending to say the problem is?

No, you got it the wrong way around. Forge's version isn't working because it checks _explicitly_ for the snow layer block. If you use it with any other block that extends BlockSnow the item is going to break...

@tterrag1098 
... and it's a kludge to have to override the method to get the item to behave as it was intended (not to mention all the poor modders who do not use forge and are likely unaware of this change in behaviour).

**Comment by tterrag1098** [comment]
There is no change in behavior. The method was added to *keep* vanilla behavior.

**Comment by Sollace** [comment]
> 
> 
> This change is still needed. Forge patches World.mayPlace to use the block's `isReplaceable` method rather than the material. This requires the check for isReplaceable to be added to ItemBlock [here](https://github.com/MinecraftForge/MinecraftForge/blame/1.12.x/patches/minecraft/net/minecraft/item/ItemBlock.java.patch#L39). Which in turn requires ItemSnow to specially override `canPlaceBlockOnSide` to handle the odd interaction logic.
> 
> TL;DR: It's intended that you need to override this function.

So you broke this item to keep from breaking other items? And you don't think it's fair to find a solution that avoids breaking _any_ items?

Well, as bs mentioned, you don't have to remove the patch entirely. You can get the same effect simply by using an instanceof check instead of an explicit equality check.

**Comment by tterrag1098** [comment]
That doesn't match vanilla's original implementation. Blame vanilla for using a hacky solution (checking the *material* replaceability in this one case only). We're just sticking to that while providing better hooks.

**Comment by Sollace** [comment]
> 
> 
> There is no change in behavior. The method was added to _keep_ vanilla behavior.

Vanilla behaviour: Right-click increases levels to a full 8-layer block, for any snowlayer type block
Current behaviour: Rick-click increases levels to a full 8-layer block, for _only_ the vanilla block. Otherwise it stops at 2 layers.


> 
> 
> That doesn't match vanilla's original implementation. Blame vanilla for using a hacky solution (checking the _material_ replaceability in this one case only). We're just sticking to that while providing better hooks.

The problem is the replaceability check isn't really what's causing the problem. It's the fact that you only do that check for one specific instance of a block type.

I'll take a wild guess and say if it was left as it was, I'd be able to extend BlockSnow and use ItemSnow just fine and it would work _perfectly_.

**Comment by kierajreed** [comment]
You aren't really supposed to extend vanilla blocks anyways... you're supposed to add the required behaviors yourself.

**Comment by Sollace** [comment]
... by extending the classes those blocks are based upon.

What I mean is:

class BlockMyLayerBlock extends BlockSnow {
   ....
}

**Comment by tterrag1098** [comment]
You guess wrong. I just spent an hour testing removing the patches to see the effects. Without the changes you cannot place blocks against the side of snow layers, because forge patches the isReplaceable call to the block which returns false. This is added compatibility, since the block variant gets more context. I'm not going to bicker with you anymore about what the hook is doing since I just spent a good amount of time figuring it out. If you have a better solution, feel free to PR it.

**Comment by Sollace** [comment]
Alright, I will try to phrase it differently:

When adding a custom layered block, right-clicking the _top_ of that block does not work. The block should increase its levels to a full 8 layers, like the snow layer does, but instead it stops at 2.

This is because of a bug in the item used for placing those types of blocks, and the only way to make them function as intended is to create a new item class and override that method as I described.

So the end result is you have to do:

public static Block slime_layer = new BlockSlimeLayer(...) // BlockSlimeLayer extends BlockSnowLayer
public static Item slime_layer = new ItemSlimeLayer(...) // ItemSlimeLayer extends ItemSnow

instead of the preferred:

public static Block slime_layer = new BlockSlimeLayer(...)
public static Item slime_layer = new ItemSnow(...)

That is all I'm trying to say.

**Comment by LexManos** [comment]
And what we are saying, is that due to the hacky nature in which Vanilla implemented this feature. The simplest and best way to fix it without breaking other vanilla mechanics, or existing mod mechanics. Was the way we did it. You should make your items only attempt to expand your target block. Not snow or other blocks.
Sorry that you have to develop for a somewhat sane API instead of the hacky things Vanilla provides. End of discussion.
